### DISCUSSION TITLE
Dice Metric and Dice Loss

### DISCUSSION THREAD

**Discussion by ajulyav** [Project-MONAI/MONAI#5699] [Q&A]
I could see the tutorials and the pipeline is the following: 

1. Compute Dice Loss on outputs sigmoid/softmax, but before post-process (which is AsDescreate) -> then this is soft Dice loss 
2. Compute Dice Metric on already post-process (i.e. binary outputs).

Though, it makes me confuse.
I could see that in this case  the Dice Loss could be very quite low, like 0.1 while Dice Metric hardly reaches 0.5 like this.

Additionally, are not they Dice Metric + Dice Loss = 1 ?  

I.e.: 

`loss = loss_function(outputs, labels)` 
is on softmax//sigmoid outputs and binary labels

```
val_outputs = [post_trans(i) for i in decollate_batch(val_outputs)]
dice_metric(y_pred=val_outputs, y=val_labels)
```
is on post-process outputs. 

So, why not compute Loss and Metric on binary, as here is pointed: https://github.com/Project-MONAI/tutorials/blob/main/modules/dice_loss_metric_notes.ipynb   -> By default the loss and metric will work on the binary case

Or, both on soft? 
In general, why all tutorials like Brats, Spleen compute Dice Score and Dice loss like that  (loss on softmax/sigmoid, score on binary)

**Comment by Joeycho** [comment]
Hi, what I understood regarding this topic, this soft concept (allowing floating numbers between 0 and 1) is for deep learning training.

If it is hardly thresholding to 0 or 1 and calculated loss, it might not be good for calculating the gradient and it ignores the difference between 0.01 and 0.4(both to 0). For me, for training purpose, soft loss makes sense, and for the metric, it should be strict since we need the exact dice score on the current prediction.

**Reply by ajulyav** [reply]
I agree that it makes sense from this point of view. Though, I am a bit confused that this way of training on soft and evaluate on binary, Dice Loss + Dice Score cannot be = 1. I just wonder if this is important or not (as everywhere I see the original Dice loss as: Dice loss = 1 - Dice).

**Reply by Joeycho** [reply]
Well, then, maybe you can compare the result between soft one and hard one. In 'Dice loss = 1 -Dice', how to calculate Dice term is slightly different among the methods. I think you just need to choose the best out of them. The Dice Loss is still motiviated from Dice Score.
